auth/src/VSCodeAzureSubscriptionProvider.ts

```diff
@@ -27,19 +27,17 @@ export class VSCodeAzureSubscriptionProvider extends vscode.Disposable implement
     private readonly onDidSignOutEmitter = new vscode.EventEmitter<void>();
     private lastSignOutEventFired: number = 0;
 
-    public constructor() {
-        const disposable = vscode.authentication.onDidChangeSessions(async e => {
-            // Ignore any sign in that isn't for the configured auth provider
-            if (e.provider.id !== getConfiguredAuthProviderId()) {
-                return;
-            }
+    private readonly sessionProvider: AzureSessionProvider;
 
-            if (await this.isSignedIn()) {
+    public constructor() {
+        const sessionProvider = new VSCodeAzureSessionProvider();

```

Comment:
Since the session provider holds state (tenants and their signed-in statuses), I'd like to avoid having multiple instantiated within the same extension context since they could become out of sync. Currently if someone does `new VSCodeAzureSessionProvider()` and `new VSCodeAzureSubscriptionProvider()`, there will be two instances.

I don't think we need to proactively prevent instantiating multiple `VSCodeSessionProvider`s, but since this might be a common scenario, I'd like to prevent it here. A simple solution would be to make the session provider a constructor argument instead.

Reply:
Agree. I was trying to avoid breaking the existing contract.

If we were going to make breaking changes, I wonder if we could go a step further and provide something like an `activateVSCodeSessionProviders` method (called by consumers from their extension's `activate` method) that creates the `VSCodeAzureXxxxProvider` instances, and registers them with the extension context (for proper disposal later, although I'm not sure how important that is). And then `getXxxxProvider` methods that return the created instances. Sort of like a singleton pattern, VS Code style.

I ended up doing something like that [here](https://github.com/Azure/vscode-aks-tools/blob/b2678a63c02add1362bd362a9c1d4d30865589ab/src/auth/azureSessionProvider.ts#L27-L36).

That might be too big of a change, but just throwing it out there for discussion.

Reply:
IMO singleton pattern here makes sense, but I think a note in our documentation and making it an optional constructor argument could work fine for now.

If we think this is a big enough issue then we can consider breaking changes.